File: String/50longestSubStrWithoutRepeatingChar/longestSubstrWithoutRepeatChar.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        subString = {}
        maxLen = 0
        curLen = 0
        l=0
        repeat = 0
        while l < len(s):
            if s[l] in subString:
                if subString[s[l]] < repeat:
                    subString[s[l]] = l
                    curLen = (l - repeat) + 1
                    maxLen = max(maxLen, curLen)
                    l += 1
                    continue
                repeat = max(repeat, subString[s[l]])
                subString[s[l]] = l
                curLen = l - repeat
                maxLen = max(maxLen, curLen)
                l += 1
                repeat += 1
                continue
            subString[s[l]] = l
            curLen = (l - repeat) + 1
            maxLen = max(maxLen, curLen)
            l += 1
        return maxLen
    def solution2(self, s: str) -> int:
        exists = {}
        l = 0
        res = 0
        for r in range(len(s)):
            if s[r] not in exists:
                res = max(res, r - l + 1)
            else:
                if exists[s[r]] < l:
                    res = max(res, r - l + 1)
                else:
                    l = exists[s[r]] + 1
            exists[s[r]] = r
        return res
print(Solution().solution2("abcabcbb"))
print(Solution().lengthOfLongestSubstring("pwwkew"))
print(Solution().lengthOfLongestSubstring("abcadefghijklmnopqrstuv"))
print(Solution().lengthOfLongestSubstring("dvdf"))
print(Solution().lengthOfLongestSubstring("aabaab!bb"))
print(Solution().lengthOfLongestSubstring("abba"))
print(Solution().lengthOfLongestSubstring("tmmzuxt"))

File: String/50longestSubStrWithoutRepeatingChar/longestSubstrWithoutRepeatChar.java

public class longestSubstrWithoutRepeatChar {
    public static int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> subString = new HashMap<>();
        int maxLen = 0,
                curLen = 0,
                l = 0,
                repeat = 0,
                strLen = s.length();
        while (l < strLen) {
            if (subString.containsKey(s.charAt(l))) {
                if (subString.get(s.charAt(l)) < repeat) {
                    subString.put(s.charAt(l), l);
                    curLen = (l - repeat) + 1;
                    maxLen = Math.max(maxLen, curLen);
                    l += 1;
                    continue;
                }
                repeat = Math.max(repeat, subString.get(s.charAt(l)));
                subString.put(s.charAt(l), l);
                curLen = l - repeat;
                maxLen = Math.max(maxLen, curLen);
                l += 1;
                repeat += 1; // 1 2 4
                continue;
            }
            subString.put(s.charAt(l), l);
            curLen = (l - repeat) + 1;
            maxLen = Math.max(maxLen, curLen);
            l += 1;
        }
        return maxLen;
    }

    public static int Sol2(String s) {
        if (s.length() == 0)
            return 0;
        HashMap<Character, Integer> exists = new HashMap<Character, Integer>();
        int res = 0;
        for (int i = 0, j = 0; i < s.length(); ++i) {
            if (exists.containsKey(s.charAt(i))) {
                j = Math.max(j, exists.get(s.charAt(i)) + 1);
            }
            exists.put(s.charAt(i), i);
            res = Math.max(res, i - j + 1);
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println(lengthOfLongestSubstring("aabaab!bb"));
    }
}